src/main.zig

```diff
@@ -1,345 +1,736 @@
-//! This module provides functions for serializing and deserializing
-//! data structures with the SSZ method.
+//! The code bellow is essentially a combination of https://github.com/Raiden1411/zabi and https://github.com/gballet/ssz.zig
+//! to the most recent version of zig with a couple of stylistic changes and support for
+//! other zig types.
 
 const std = @import("std");
+const testing = std.testing;
+const Allocator = std.mem.Allocator;
 const ArrayList = std.ArrayList;
-const builtin = std.builtin;
 const sha256 = std.crypto.hash.sha2.Sha256;
-const hashes_of_zero = @import("./zeros.zig").hashes_of_zero;
-const Allocator = std.mem.Allocator;
 
-/// Number of bytes per chunk.
-const BYTES_PER_CHUNK = 32;
+/// Set of possible errors while performing ssz decoding.
+pub const SSZDecodeErrors = Allocator.Error || error{ InvalidEnumType, IndexOutOfBounds };
 
-/// Number of bytes per serialized length offset.
-const BYTES_PER_LENGTH_OFFSET = 4;
-
-// Determine the serialized size of an object so that
-// the code serializing of variable-size objects can
-// determine the offset to the next object.
-fn serializedSize(comptime T: type, data: T) !usize {
+/// Performs ssz decoding according to the [specification](https://ethereum.org/developers/docs/data-structures-and-encoding/ssz).
+pub fn decodeSSZ(comptime T: type, serialized: []const u8) SSZDecodeErrors!T {
     const info = @typeInfo(T);
-    return switch (info) {
-        .Array => data.len,
-        .Pointer => switch (info.Pointer.size) {
-            .Slice => data.len,
-            else => serializedSize(info.Pointer.child, data.*),
-        },
-        .Optional => if (data == null)
-            @as(usize, 1)
-        else
-            1 + try serializedSize(info.Optional.child, data.?),
-        .Null => @as(usize, 0),
-        else => error.NoSerializedSizeAvailable,
-    };
-}
 
-/// Returns true if an object is of fixed size
-fn isFixedSizeObject(comptime T: type) !bool {
-    const info = @typeInfo(T);
     switch (info) {
-        .Bool, .Int, .Null => return true,
-        .Array => return false,
-        .Struct => inline for (info.Struct.fields) |field| {
-            if (!try isFixedSizeObject(field.type)) {
-                return false;
-            }
+        .bool => return serialized[0] != 0,
+        .int => |int_info| return std.mem.readInt(T, serialized[0..@divExact(int_info.bits, 8)], .little),
+        .optional => |opt_info| {
+            if (serialized.len > 0) {
```

Comment:
This was one change I had to make; serialization was failing on serializing one of the optionals.

Reply:
Which case will `serialized.len = 0`? Have you add some test for the case?

Reply:
`serialized.len == 0` for when an optional is being deserialized. Specifically, see this [test](https://github.com/zen-eth/ssz.zig/pull/2/files#diff-bd1a1eaede544f87e2ad772fcf97bc3d6b01ce5b15ed2a54a5b76097ff4aa242R223)

But looking at the spec again, it seems like `null` is an illegal type except for when in the first value of a union, so i'm questioning if this decoding should even be possible.

Reply:
I see the encode method, of it's null, it will write 0. So I think even if no `serialized.len > 0`, the test will pass too.
```
        .optional => {
            if (value) |val| {
                try writer.writeInt(u8, 1, .little);
                return try encodeItem(val, list);
            } else try writer.writeInt(u8, 0, .little);
        },
```
The ssz `union` type may have multi type, the first byte maybe 0,1,2..., the `optional` can only present the 0 and 1.

Reply:
The error comes during deserializing. It tries to index into the first value of the slice (which is empty)

Reply:
I think it not the problem of `optional`, it's a problem for all types.  if `serialized` is empty, other types like `.bool => return serialized[0] != 0` will have the same problem. maybe the check should do at the start of `decodeSSZ` method

Reply:
@fearlessfe good point! Since this PR is mainly about porting zephyrus' ssz impl here (and we've accomplished that with unit tests passing even if they might not be the most comprehensive tests), let's save that for a followup PR and get this merged - what do you think?